src/nvim/api/vim.c

```diff
@@ -2389,3 +2389,169 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
+    // Just redraw the 'winbar' or 'statusline'.
+    win_check_ns_hl(wp);
+    if (status & kStatusBar) {
+      win_redr_winbar(wp);
+    }
+    if (status & kStatusLine) {
+      win_redr_status(wp);
+    }
+    win_check_ns_hl(NULL);
+  }
+}
+
+/// Instruct Nvim to redraw various elements.
+///
+/// @param opts  Optional parameters.
+///               - flush: Update the screen with pending updates, even if
+///                 Nvim is waiting for input and would normally not do so.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows when both are omitted.
+///               - range: Buffer range to redraw, useful for decoration
+///                 providers. Tuple `[first, last]`:
+///                   - first: first line number of the range
+///                   - last: last line number of the range
+///               - cursor: Immediately update cursor position on the screen.
+///               - status: Redraws (a combination of) the following status
+///                 components (array of one or more strings):
+///                 - "bar": |'winbar'|
+///                 - "column": |'statuscolumn'|
+///                 - "line": |'statusline'|
+///                 - "tab": |'tabline'|
+///                 - "all": All of the above
+///               - win: Target a specific |window-ID| for `cursor`, `status`
+///                 and `clear` updates. Current window when 0, all windows if
+///                 omitted and applicable.
+///               - buf: Target a specific buffer number for `range`, `status`,
+///                 and `clear` updates, current buffer when 0 or omitted.
+void nvim_redraw(Dict(redraw) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  win_T *win = NULL;
+  buf_T *buf = NULL;
+  bool save_lz = p_lz;
+  int save_rd = RedrawingDisabled;
+  RedrawingDisabled = 0;
+  p_lz = false;
+
+  if (HAS_KEY(opts, redraw, win)) {
+    win = find_window_by_handle(opts->win, err);
+    if (ERROR_SET(err)) {
+      goto theend;
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, buf)) {
+    buf = find_buffer_by_handle(opts->buf, err);
+    if (ERROR_SET(err)) {
+      goto theend;
+    }
+  }
+
+  if (opts->clear) {
+    if (win != NULL) {
+      redraw_later(win, UPD_NOT_VALID);
+    } else if (buf != NULL) {
+      redraw_buf_later(buf, UPD_NOT_VALID);
+    } else {
+      redraw_all_later(UPD_NOT_VALID);
+    }
+  }
+
+  if (HAS_KEY(opts, redraw, range)) {
+    VALIDATE(kv_size(opts->range) == 2
+             && kv_A(opts->range, 0).type == kObjectTypeInteger
+             && kv_A(opts->range, 1).type == kObjectTypeInteger,
+             "%s", "Invalid 'range': expected 2-tuple of Lua numbers", {
+      goto theend;
+    });
+    Integer first = kv_A(opts->range, 0).data.integer;
+    Integer last = kv_A(opts->range, 1).data.integer;
+    nvim__buf_redraw_range(buf ? buf->handle : curbuf->handle, first, last, err);
```

Comment:
then, if `win` is supplied but not `buf`, this either needs to throw an error or use the buffer of that window (ideally it would imply only redrawing the range for the specific window, but doing it for the buffer in the window is good enough).

Reply:
Yes, I considered exposing `win_upate()` to be able to target specific windows but I'm not sure that's a good idea. It would only be effective when `redrawing() == false` anyways, otherwise the rest of the windows would be updated at the next `update_screen()` anyways.